### PR TITLE
Support Python 3

### PR DESCRIPTION
**Pull request by siebeclaes**

### CODE REVIEW

**Comment by siebeclaes** [comment]
@pdellaert 
I added a commit to fix support for multiple sessions.

You should always call start() on every session you want to use in the script before using a with statement. Following snippet works both using Python 2.7 and 3:

    session1 = vsdk.NUVSDSession( username=username, password=password, 
            enterprise=enterprise, api_url="https://10.30.31.73:8443")
    session2 = vsdk.NUVSDSession( username=username, password=password, 
            enterprise=enterprise, api_url="https://10.30.31.60:8443")

    session1.start()
    session2.start()

    with session1 as s:
        ent = vsdk.NUEnterprise(name="p3-Level1")
        s.user.create_child(ent)
        
        with session2 as s:
            ent = vsdk.NUEnterprise(name="p3-Level2")
            s.user.create_child(ent)
        
            with session1 as s:
                ent = vsdk.NUEnterprise(name="p3-Level3")
                s.user.create_child(ent)
        
                with session2 as s:
                    ent = vsdk.NUEnterprise(name="p3-Level4")
                    s.user.create_child(ent)

Outside a with statement, vspk will use the last session start() was called on. In the snippet, actions on session2 will still work but using session1 might result in undefined behavior. I may be able to include an extra check that stops the user from doing this but I haven't looked at it yet.

File: bambou/nurest_object.py [hunk 1]
```diff
@@ -122,14 +124,14 @@ def _compute_args(self, data=dict(), **kwargs):
                 kwargs: a list of arguments
         """
 
-        for name, remote_attribute in self._attributes.items():
+        for name, remote_attribute in list(self._attributes.items()):
```
Comment (little-dude) [inline bambou/nurest_object.py]: I don't think you need to make this a list if you just iterate

File: bambou/nurest_object.py [hunk 2]
```diff
@@ -352,7 +354,7 @@ def validate(self):
                 continue  # without error
 
             if type(value) != attribute.attribute_type:
-                if attribute.attribute_type != str or type(value) != unicode:
+                if attribute.attribute_type != str or type(value) != str:
```
Comment (little-dude) [inline bambou/nurest_object.py]: I'm not sure that does what we expect in python 2. `value` is provided by users, will most likely be a python2 `str`. So `type(value)` will always be different than `builtins.str`. Unless you convert users input in the attribute setter?
  
```
In [3]: def is_text(value):          
   ...:     from builtins import str 
   ...:     return type(value) == str
   ...:                              
   ...:                              
                                     
In [4]: is_text(str('coucou'))       
Out[4]: False                        
```

File: bambou/nurest_session.py [hunk 3]
```diff
@@ -146,32 +152,21 @@ def _authenticate(self):
...
 
-        try:
-            frame = inspect.stack()[1][0]
-        except IndexError:
-            _NURESTSessionCurrentContext.session = self
-            self._authenticate()
-            return self
+        if NURESTSession.session_stack:
+            bambou_logger.critical("VSPK currently does not support starting a session using start() inside a with statement. Please open all sessions outside the outermost with statement. Exiting for safety")
+            sys.exit(-1)
```
Comment (little-dude) [inline bambou/nurest_session.py]: Why not raising an exception here instead?

File: bambou/nurest_connection.py [hunk 4]
```diff
@@ -338,8 +339,11 @@ def _did_timeout(self):
     def _make_request(self, session=None):
         """ Make a synchronous request """
 
-        from .nurest_session import _NURESTSessionCurrentContext
-        _NURESTSessionCurrentContext.session = session
+        ### from .nurest_session import NURESTSession
```
Comment (little-dude) [inline bambou/nurest_connection.py]: This can be removed completely no?

**Review by little-dude — COMMENTED** [review]
This is really nice :) I left some comments about strings.

File: bambou/nurest_object.py (hunk 2 above)
Comment (pdellaert) [reply bambou/nurest_object.py]: Looking at this, i'm pretty sure the validation is good:
https://gist.github.com/pdellaert/026ab1da148f301935af436a3eed145c
(and yes, enterprise.name is a str type attribute)

File: bambou/nurest_session.py (hunk 3 above)
Comment (pdellaert) [reply bambou/nurest_session.py]: agree

File: bambou/nurest_object.py (hunk 1 above)
Comment (pdellaert) [reply bambou/nurest_object.py]: correct

File: bambou/nurest_connection.py (hunk 4 above)
Comment (pdellaert) [reply bambou/nurest_connection.py]: yes

**Comment by pdellaert** [comment]
@siebeclaes, i've added 4 commits to handle Corentin's comments and do some cleanup.

Tests I've done, all done both with Python 2.7.13 and Python 3.6.3:
- Push channel
- Certificate based auth
- Password based auth
- `get` - normal and with filter and paging
- `get_first` - normal and with filter
- `fetch`
- `create_child`
- `assign`
- `save`
- `delete`
- `validate`
- Multi-VSD sessions

Everything looks good to me

Also important: VSPK requirements for bambou are <3.0, so when we release this, by default, installing VSPK will not use bambou 3.0.0 (Python3 support). 

I suggest we release Bambou 3.0.0, leave VSPK as is until 5.2.2, at which time we update the requirements of VSPK, so it also installs Bambou 3.0.0. This allows us to give a procedure to Nuage field teams and customers on how to enable Python 3 support (basically installing vspk and installing/upgrading to a specific version of bambou)

@little-dude , @siebeclaes , let me know if you are both happy, i'll merge, tag and release a new build on pypi

File: bambou/nurest_object.py (hunk 2 above)
Comment (siebeclaes) [reply bambou/nurest_object.py]: @little-dude is right. The validation can fail with the current code. I have an idea on how to fix it and still be python2/3 compatible. Will provide a commit tomorrow

Comment (pdellaert) [reply bambou/nurest_object.py]: Cool, thanks!

**Comment by siebeclaes** [comment]
@pdellaert @little-dude 
I removed `from builtins import str` from `nurest_object.py` so on python 2, when an attribute's type is set to str, it is a python 2 str. And I added the unicode check again but only on python 2 so we still accept values where `type(value) == unicode`.

**Comment by siebeclaes** [comment]
I did a simple test to add an enterprise on python 2 and 3. On python 2, I tested with name as str and unicode and both worked. Python 3 also works.

I think we're good to release now

**Comment by pdellaert** [comment]
@siebeclaes , in Python 3, we now are no longer checking if it is a proper `str`?

**Comment by siebeclaes** [comment]
We are, we define attribute_type as str in the vspk objects, so the the if statement just above the one I changed checks if the attribute_type corresponds to `type(value)` which on python 3 checks if you actually provided a str if we expect one

**Comment by pdellaert** [comment]
Right! I must have missed that, sorry

**Comment by little-dude** [comment]
Do you mind if we hold off release for a few days? We have a number of tools and regressions using vspk internally. I'd like to run those with this PR just to be 100% we're not breaking anything.

**Comment by pdellaert** [comment]
@little-dude sure, but as i said: My plan is to release this as version 3.0.0. 

VSPK-Python has a requirement: `bambou>=2.0,<3.0`
As a result, a default install of VSPK Python from pip will **not** install this new version, it will install the previous version. 

I'm going to wait until 5.2.2 to actually change that requirement on VSPK-Python, to give people the time to actually test this. The procedure to test is simple, just do:  
```
pip install vspk
pip uninstall bambou
git clone https://github.com/nuagenetworks/bambou
cd bambou
pip install .
```

That gives a lot of time for testing.

**Comment by siebeclaes** [comment]
Just a sidenote: you can just do `pip install --upgrade https://github.com/nuagenetworks/bambou` as well

**Comment by pdellaert** [comment]
@little-dude have you made progress on that testing?

**Pull request merged**